nestjs/src/listeners/course.listener.ts

```diff
@@ -0,0 +1,118 @@
+import { S3 } from '@aws-sdk/client-s3';
+import { Course } from '@entities/course';
+import { HttpService } from '@nestjs/axios';
+import { Injectable, Logger } from '@nestjs/common';
+import { Cron, CronExpression } from '@nestjs/schedule';
+import { InjectRepository } from '@nestjs/typeorm';
+import { ExportCourseDto } from 'src/courses/dto';
+import { FindOptionsWhere, MoreThanOrEqual, Repository } from 'typeorm';
+import { ConfigService } from '../config';
+
+@Injectable()
+export class CourseListener {
+  private s3: S3;
+  private logger = new Logger(CourseListener.name);
+
+  private readonly objectKey = 'app/courses.json';
+  private readonly bucket;
+
+  constructor(
+    @InjectRepository(Course)
+    private readonly repository: Repository<Course>,
+    private readonly httpService: HttpService,
+    readonly configService: ConfigService,
+  ) {
+    this.s3 = new S3(this.configService.awsClient);
+    this.bucket = this.configService.buckets.cdn;
+  }
+
+  @Cron(CronExpression.EVERY_30_MINUTES)
+  async handleCoursesChange() {
+    // check if courses are changed since last export
+    const changed = await this.areCoursesChanged();
+    if (!changed) {
+      return;
+    }
+
+    // export courses to cdn
+    await this.exportCoursesToCdn();
+
+    // trigger rs site update
+    this.triggerSiteUpdate();
+  }
+
+  /**
+   * Check if courses are changed since last export
+   * @returns true if courses are changed, false otherwise
+   */
+  private async areCoursesChanged() {
+    const response = await this.s3.headObject({ Bucket: this.bucket, Key: this.objectKey });
+    const lastModifiedCourse = await this.repository.findOne({
+      where: this.getCoursesWhereFilter(),
+      order: { updatedDate: 'DESC' },
+    });
```

Comment:
As I can see these API calls don't depend on each other  
```suggestion
    const [response, lastModifiedCourse] = await Promise.all([
      this.s3.headObject({ Bucket: this.bucket, Key: this.objectKey }),
      this.repository.findOne({
        where: this.getCoursesWhereFilter(),
        order: { updatedDate: 'DESC' },
      }),
    ]);
```

Reply:
not actual anymore.